File: backend/app/schemas/studio.py
"""内容工坊模块 Pydantic Schema。"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.sensitive_filter import contains_blocked


def _check_no_blocked(v: str | None) -> str | None:
    if v is not None and contains_blocked(v):
        raise ValueError("文本包含敏感词")
    return v


# ============================================================
# 项目
# ============================================================


class StudioProjectCreate(BaseModel):
    shop_id: uuid.UUID
    title: str = Field(..., min_length=1, max_length=100)

    @field_validator("title")
    @classmethod
    def title_blocked(cls, v: str) -> str:
        if contains_blocked(v):
            raise ValueError("文本包含敏感词")
        return v


class StudioProjectUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=100)
    status: Literal["draft", "generated"] | None = None

    @field_validator("title")
    @classmethod
    def title_blocked(cls, v: str | None) -> str | None:
        return _check_no_blocked(v)


class StudioProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    shop_id: uuid.UUID
    title: str
    status: str
    created_at: datetime
    updated_at: datetime


class StudioCopySummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    input_payload: dict | None = None
    titles: list | None = None
    body: str | None = None
    tags: list | None = None
    image_guide: dict | None = None
    created_at: datetime


class StudioDeckSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    copy_id: uuid.UUID
    template: str
    theme: str
    page_count: int
    page_specs: list | None = None
    source_assets: list | None = None
    images: list | None = None
    qa_report: dict | None = None
    status: str
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class StudioProjectDetail(StudioProjectResponse):
    copies: list[StudioCopySummary] = []
    decks: list[StudioDeckSummary] = []


# ============================================================
# 文案
# ============================================================


class CopyGenerateRequest(BaseModel):
    category: str = Field(..., min_length=1, max_length=50)
    style: str = Field(..., min_length=1, max_length=50)
    price_range: str = Field(..., min_length=1, max_length=50)
    topic: str = Field(..., min_length=1, max_length=200)
    shop_name: str = Field(..., min_length=1, max_length=100)

    @field_validator("category", "style", "price_range", "topic", "shop_name")
    @classmethod
    def no_blocked(cls, v: str) -> str:
        if contains_blocked(v):
            raise ValueError("文本包含敏感词")
        return v


class CopyTitleItem(BaseModel):
    text: str = Field(..., max_length=50)
    strategy: str = Field(..., max_length=50)


class ImageGuidePage(BaseModel):
    position: str = Field(..., max_length=100)
    purpose: str = Field(..., max_length=100)
    prompt: str = Field(..., max_length=2000)


class CopyImageGuide(BaseModel):
    cover_prompt: str = Field(..., max_length=2000)
    pages: list[ImageGuidePage] = []


class CopyGenerateResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    input_payload: dict
    titles: list[CopyTitleItem]
    body: str
    tags: list[str]
    image_guide: CopyImageGuide
    created_at: datetime


class CopyUpdateRequest(BaseModel):
    titles: list[CopyTitleItem] | None = None
    body: str | None = Field(None, min_length=1, max_length=5000)
    tags: list[str] | None = None
    image_guide: CopyImageGuide | None = None

    @field_validator("body")
    @classmethod
    def body_blocked(cls, v: str | None) -> str | None:
        return _check_no_blocked(v)


# ============================================================
# 卡组
# ============================================================


class DeckCreateRequest(BaseModel):
    copy_id: uuid.UUID
    template: Literal["editorial", "swiss"]
    theme: str = Field(..., min_length=1, max_length=50)
    page_count: int
    asset_ids: list[uuid.UUID] = Field(default_factory=list, max_length=8)


class DeckCreateResponse(BaseModel):
    deck_id: uuid.UUID
    images: list[dict] = []
    qa_report: dict | None = None
    status: str
    error_message: str | None = None


class ExportToDesignResponse(BaseModel):
    design_project_id: uuid.UUID
    asset_ids: list[uuid.UUID]
# ============================================================
# 配图提示词丰富
# ============================================================


class ImagePromptEnrichRequest(BaseModel):
    direction: str = Field(..., min_length=1, max_length=500)

    @field_validator("direction")
    @classmethod
    def direction_blocked(cls, v: str) -> str:
        if contains_blocked(v):
            raise ValueError("文本包含敏感词")
        return v


class ImagePromptEnrichResponse(BaseModel):
    main_idea: str
    prompt: str

File: frontend/src/services/studio.ts
import api from "./api";

export type StudioProjectStatus = "draft" | "generated";
export type DeckTemplate = "editorial" | "swiss";
export type DeckStatus = "draft" | "rendered" | "failed";

export interface StudioProject {
  id: string;
  shop_id: string;
  title: string;
  status: StudioProjectStatus;
  created_at: string;
  updated_at: string;
}

export interface CopyTitleItem {
  text: string;
  strategy: string;
}

export interface ImageGuidePage {
  position: string;
  purpose: string;
  prompt: string;
}

export interface CopyImageGuide {
  cover_prompt: string;
  pages: ImageGuidePage[];
}

export interface StudioCopy {
  id: string;
  project_id: string;
  input_payload: Record<string, string> | null;
  titles: CopyTitleItem[] | null;
  body: string | null;
  tags: string[] | null;
  image_guide: CopyImageGuide | null;
  created_at: string;
}

export interface PageSpec {
  title: string;
  bullets: string[];
  image_index: number | null;
}

export interface DeckImage {
  page: number;
  url: string;
  width: number;
  height: number;
}

export interface SourceAsset {
  source: "design" | "upload";
  asset_id?: string | null;
  url?: string | null;
}

export interface QaPage {
  page: number;
  pass: boolean;
  checks: {
    density: { pass: boolean; coverage: number; bands: number[]; issues: string[] };
    overflow: { pass: boolean; overflow_px: number };
    bottom_blank: { pass: boolean; bottom_gap_px: number };
  };
  issues: string[];
}

export interface QaReport {
  all_pass: boolean;
  pages: QaPage[];
}

export interface StudioDeck {
  id: string;
  project_id: string;
  copy_id: string;
  template: DeckTemplate;
  theme: string;
  page_count: number;
  page_specs: PageSpec[] | null;
  source_assets: SourceAsset[] | null;
  images: DeckImage[] | null;
  qa_report: QaReport | null;
  status: DeckStatus;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface StudioProjectDetail extends StudioProject {
  copies: StudioCopy[];
  decks: StudioDeck[];
}

export interface CopyGenerateRequest {
  category: string;
  style: string;
  price_range: string;
  topic: string;
  shop_name: string;
}

export interface DeckCreateRequest {
  copy_id: string;
  template: DeckTemplate;
  theme: string;
  page_count: number;
  asset_ids: string[];
}

export interface DeckCreateResponse {
  deck_id: string;
  images: DeckImage[];
  qa_report: QaReport | null;
  status: DeckStatus;
  error_message: string | null;
}

export interface ExportToDesignResponse {
  design_project_id: string;
  asset_ids: string[];
}

export const studioService = {
  // ---- projects ----
  listProjects: (shopId?: string) =>
    api.get<StudioProject[]>("/studio/projects", { params: { shop_id: shopId } }),

  createProject: (data: { shop_id: string; title: string }) =>
    api.post<StudioProject>("/studio/projects", data),

  getProject: (id: string) => api.get<StudioProjectDetail>(`/studio/projects/${id}`),

  updateProject: (id: string, data: { title?: string; status?: StudioProjectStatus }) =>
    api.patch<StudioProject>(`/studio/projects/${id}`, data),

  deleteProject: (id: string) => api.delete<{ ok: boolean }>(`/studio/projects/${id}`),

  // ---- copies ----
  generateCopy: (projectId: string, data: CopyGenerateRequest) =>
    api.post<StudioCopy>(`/studio/projects/${projectId}/copy/generate`, data, {
      timeout: 60_000,
    }),

  updateCopy: (copyId: string, data: { titles?: CopyTitleItem[]; body?: string; tags?: string[]; image_guide?: CopyImageGuide }) =>
    api.patch<StudioCopy>(`/studio/copies/${copyId}`, data),

  enrichImagePrompt: (copyId: string, direction: string) =>
    api.post<{ main_idea: string; prompt: string }>(
      `/studio/copies/${copyId}/image-prompt/enrich`,
      { direction },
      { timeout: 60_000 }
    ),

  // ---- decks ----
  createDeck: (projectId: string, body: DeckCreateRequest) =>
    api.post<DeckCreateResponse>(`/studio/projects/${projectId}/decks`, body, {
      timeout: 180_000,
    }),

  createDeckWithFiles: (
    projectId: string,
    params: {
      copy_id: string;
      template: DeckTemplate;
      theme: string;
      page_count: number;
      asset_ids: string[];
      files: File[];
    }
  ) => {
    const fd = new FormData();
    fd.append("copy_id", params.copy_id);
    fd.append("template", params.template);
    fd.append("theme", params.theme);
    fd.append("page_count", String(params.page_count));
    if (params.asset_ids.length > 0) {
      fd.append("asset_ids", JSON.stringify(params.asset_ids));
    }
    params.files.forEach((f) => fd.append("files", f));
    return api.post<DeckCreateResponse>(`/studio/projects/${projectId}/decks`, fd, {
      timeout: 180_000,
    });
  },

  getDeck: (deckId: string) => api.get<StudioDeck>(`/studio/decks/${deckId}`),

  exportToDesign: (deckId: string) =>
    api.post<ExportToDesignResponse>(`/studio/decks/${deckId}/export-to-design`),
};
